File: server/controllers/pageController.js
const Page = require('../models/Page.js');

// Get all pages
const getAllPages = async (req, res) => {
    try {
        const pages = await Page.find({ isPublished: true });
        res.status(200).json({ pages, success: true });
    } catch (err) {
        res.status(500).json({ message: 'Error fetching pages', error: err.message });
    }
};

// Get single page by slug (public)
const getPage = async (req, res) => {
    try {
        const { slug } = req.params;
        const page = await Page.findOne({ slug });
        if (!page) return res.status(404).json({ message: 'Page not found' });
        res.status(200).json(page);
    } catch (err) {
        res.status(500).json({ message: 'Error fetching page', error: err.message });
    }
};

// Get single page by ID (admin)
const getPageById = async (req, res) => {
    try {
        const { id } = req.params;
        const page = await Page.findById(id);
        if (!page) return res.status(404).json({ message: 'Page not found' });
        res.status(200).json(page);
    } catch (err) {
        res.status(500).json({ message: 'Error fetching page', error: err.message });
    }
};

// Create page (admin)
const createPage = async (req, res) => {
    try {
        const { title, slug, description, content, metaTitle, metaDescription, metaKeywords, isPublished } = req.body;
        
        // Check if page with same title or slug already exists
        const existingPage = await Page.findOne({
            $or: [
                { title: title },
                { slug: slug || title.toLowerCase().replace(/\s+/g, '-') }
            ]
        });
        
        if (existingPage) {
            return res.status(400).json({ 
                message: 'A page with this title or slug already exists',
                error: 'Duplicate page'
            });
        }
        
        const newPage = new Page({
            title,
            slug: slug || title.toLowerCase().replace(/\s+/g, '-'),
            description,
            content,
            metaTitle: metaTitle || title || '',
            metaDescription,
            metaKeywords,
            isPublished: isPublished !== undefined ? isPublished : true
        });
        
        await newPage.save();
        res.status(201).json({ message: 'Page created successfully', page: newPage });
    } catch (err) {
        console.error('Error creating page:', err);
        res.status(500).json({ message: 'Error creating page', error: err.message });
    }
};

// Update page (admin)
const updatePage = async (req, res) => {
    try {
        const { id } = req.params;
        const updates = req.body;
        
        // If updating title or slug, check for duplicates (but exclude current page)
        if (updates.title || updates.slug) {
            const existingPage = await Page.findOne({
                $or: [
                    { title: updates.title },
                    { slug: updates.slug }
                ],
                _id: { $ne: id } // Exclude current page
            });
            
            if (existingPage) {
                return res.status(400).json({ 
                    message: 'A page with this title or slug already exists',
                    error: 'Duplicate page'
                });
            }
        }
        
        const page = await Page.findByIdAndUpdate(id, updates, { new: true });
        if (!page) return res.status(404).json({ message: 'Page not found' });
        
        res.status(200).json({ message: 'Page updated successfully', page });
    } catch (err) {
        console.error('Error updating page:', err);
        res.status(500).json({ message: 'Error updating page', error: err.message });
    }
};

// Delete page (admin)
const deletePage = async (req, res) => {
    try {
        const { id } = req.params;
        const page = await Page.findByIdAndDelete(id);
        if (!page) return res.status(404).json({ message: 'Page not found' });
        
        res.status(200).json({ message: 'Page deleted successfully' });
    } catch (err) {
        res.status(500).json({ message: 'Error deleting page', error: err.message });
    }
};

module.exports = {
    getAllPages,
    getPage,
    getPageById,
    createPage,
    updatePage,
    deletePage
};

File: server/controllers/page_controller.py
from pymongo import MongoClient
import os
from bson import ObjectId
from datetime import datetime
from starlette.concurrency import run_in_threadpool

client = MongoClient(os.getenv("DATABASE_URL", "mongodb://localhost:27017"))
db = client[os.getenv("MONGO_DB_NAME", "ecommerce")]

def _serialize(doc: dict):
    if doc is None:
        return None
    def _walk(o):
        from bson import ObjectId
        if isinstance(o, ObjectId):
            return str(o)
        if isinstance(o, datetime):
            return o.isoformat()
        if isinstance(o, dict):
            return {k: _walk(v) for k, v in o.items()}
        if isinstance(o, list):
            return [_walk(i) for i in o]
        return o
    return _walk(doc)

async def get_pages():
    try:
        print("[page_controller] get_pages() called")
        def _fetch():
            print("[page_controller] _fetch() - getting pages collection")
            coll = db.get_collection("pages")
            print("[page_controller] _fetch() - executing find()")
            cursor = coll.find({}).sort("createdAt", -1)
            print("[page_controller] _fetch() - serializing pages")
            pages = [_serialize(p) for p in cursor]
            print(f"[page_controller] _fetch() - found {len(pages)} pages")
            return pages
        
        print("[page_controller] get_pages() - calling run_in_threadpool")
        pages = await run_in_threadpool(_fetch)
        print(f"[page_controller] get_pages() - returning {len(pages)} pages")
        return pages
    except Exception as e:
        print(f"[page_controller] get_pages error: {type(e).__name__}: {repr(e)}")
        import traceback
        traceback.print_exc()
        raise

async def get_page_by_slug(slug: str):
    try:
        def _fetch():
            coll = db.get_collection("pages")
            page = coll.find_one({"slug": slug})
            return page
        
        page = await run_in_threadpool(_fetch)
        return _serialize(page) if page else None
    except Exception as e:
        print("page_controller.get_page_by_slug error:", type(e).__name__, repr(e))
        import traceback
        traceback.print_exc()
        return None

async def get_page_by_id(page_id: str):
    try:
        def _fetch():
            coll = db.get_collection("pages")
            page = coll.find_one({"_id": ObjectId(page_id)})
            return page
        
        page = await run_in_threadpool(_fetch)
        return _serialize(page) if page else None
    except Exception as e:
        print("page_controller.get_page_by_id error:", type(e).__name__, repr(e))
        import traceback
        traceback.print_exc()
        return None

async def create_page(data: dict, user=None):
    try:
        def _create():
            coll = db.get_collection("pages")
            page_data = {
                "title": data.get("title"),
                "slug": data.get("slug") or (data.get("title") or "").lower().replace(" ", "-") ,
                "description": data.get("description"),
                "content": data.get("content"),
                "metaTitle": data.get("metaTitle"),
                "metaDescription": data.get("metaDescription"),
                "metaKeywords": data.get("metaKeywords"),
                "isPublished": data.get("isPublished", True),
                "createdAt": datetime.utcnow(),
            }
            result = coll.insert_one(page_data)
            page = coll.find_one({"_id": result.inserted_id})
            return page
        
        page = await run_in_threadpool(_create)
        return _serialize(page)
    except Exception as e:
        print("page_controller.create_page error:", type(e).__name__, repr(e))
        import traceback
        traceback.print_exc()
        raise

async def update_page(page_id: str, data: dict, user=None):
    try:
        def _update():
            coll = db.get_collection("pages")
            update_data = {
                "title": data.get("title"),
                "slug": data.get("slug"),
                "description": data.get("description"),
                "content": data.get("content"),
                "metaTitle": data.get("metaTitle"),
                "metaDescription": data.get("metaDescription"),
                "metaKeywords": data.get("metaKeywords"),
                "isPublished": data.get("isPublished", True),
                "updatedAt": datetime.utcnow(),
            }
            # remove None values
            update_data = {k: v for k, v in update_data.items() if v is not None}
            result = coll.update_one({"_id": ObjectId(page_id)}, {"$set": update_data})
            if result.matched_count == 0:
                raise Exception("Page not found")
            page = coll.find_one({"_id": ObjectId(page_id)})
            return page
        
        page = await run_in_threadpool(_update)
        return _serialize(page)
    except Exception as e:
        print("page_controller.update_page error:", type(e).__name__, repr(e))
        import traceback
        traceback.print_exc()
        raise

async def delete_page(page_id: str, user=None):
    try:
        def _delete():
            coll = db.get_collection("pages")
            result = coll.delete_one({"_id": ObjectId(page_id)})
            if result.deleted_count == 0:
                raise Exception("Page not found")
            return {"message": "Page deleted successfully"}
        
        result = await run_in_threadpool(_delete)
        return result
    except Exception as e:
        print("page_controller.delete_page error:", type(e).__name__, repr(e))
        import traceback
        traceback.print_exc()
        raise
